### PR TITLE
Read serial

### PR DESCRIPTION
**Pull request by abrauchli**
Add functions to read out the serial number for SHT3x/SHT85 and SHTC1 series (SHTC1, SHTC3, SHTW1, SHTW2)

### CODE REVIEW

File: sht3x/sht3x.c [hunk 1]
```diff
@@ -94,16 +95,27 @@ int16_t sht3x_read(int32_t *temperature, int32_t *humidity) {
...
 }
 
+int16_t sht3x_read_serial(uint32_t *serial) {
+    int16_t ret;
+
+    ret = sensirion_i2c_delayed_read_cmd(
+        SHT3X_ADDRESS, SHT3X_CMD_READ_SERIAL_ID, SHT3X_CMD_DURATION_USEC,
+        (uint16_t *)serial, SENSIRION_NUM_WORDS(*serial));
+    SENSIRION_WORDS_TO_BYTES(serial, SENSIRION_NUM_WORDS(*serial));
+    *serial = be32_to_cpu(*serial);
```
Comment (rnestler) [inline sht3x/sht3x.c]: These conversions from bytes to u16, to bytes and then to u32 seems a bit strange. Should we add a read bytes version to embedded-common? We have this at several places, right? Also the constant swapping makes the code hard to understand and verify.

Maybe just read to a `u16t serial_number_words[2]` and then `*serial =(serial_number_words[0]<<16) | serial_number_words[1];`, which avoids some of the byte shuffling.

File: shtc1/shtc1.c [hunk 2]
```diff
@@ -203,6 +203,34 @@ void shtc1_enable_low_power_mode(uint8_t enable_low_power_mode) {
         enable_low_power_mode ? SHTC1_CMD_MEASURE_LPM : SHTC1_CMD_MEASURE_HPM;
 }
 
+int16_t shtc1_read_serial(uint32_t *serial) {
+    int16_t ret;
+    const uint16_t tx_words[] = {0x007B};
+    uint16_t *serial_words = (uint16_t *)serial;
+
+    PM_WAKE(ret,
+            sensirion_i2c_write_cmd_with_args(SHTC1_ADDRESS, 0xC595, tx_words,
```
Comment (rnestler) [inline shtc1/shtc1.c]: Why the magic constant here instead of a `#define`?

File: shtc1/shtc1.c [hunk 3]
```diff
@@ -203,6 +203,34 @@ void shtc1_enable_low_power_mode(uint8_t enable_low_power_mode) {
         enable_low_power_mode ? SHTC1_CMD_MEASURE_LPM : SHTC1_CMD_MEASURE_HPM;
 }
 
+int16_t shtc1_read_serial(uint32_t *serial) {
+    int16_t ret;
+    const uint16_t tx_words[] = {0x007B};
+    uint16_t *serial_words = (uint16_t *)serial;
```
Comment (rnestler) [inline shtc1/shtc1.c]: I'm not sure if we hurt strict-aliasing rules here. Could we add `-Wstrict-aliasing=1` to the Makefile? Also maybe just use a temporary uint16_t serial_words[2] array.

File: shtc1/shtc1.c [hunk 4]
```diff
@@ -203,6 +203,34 @@ void shtc1_enable_low_power_mode(uint8_t enable_low_power_mode) {
...
+        SHTC1_ADDRESS, 0xC7F7, SHTC1_CMD_DURATION_USEC, &serial_words[0], 1);
+    if (ret)
+        return PM_SLEEP(ret);
+
+    ret = sensirion_i2c_delayed_read_cmd(
+        SHTC1_ADDRESS, 0xC7F7, SHTC1_CMD_DURATION_USEC, &serial_words[1], 1);
+    if (ret)
+        return PM_SLEEP(ret);
+
+    SENSIRION_WORDS_TO_BYTES(serial, SENSIRION_NUM_WORDS(*serial));
```
Comment (rnestler) [inline shtc1/shtc1.c]: shouldn't this use `serial_words`? Probably also better to just do `*serial =(serial_number_words[0]<<16) | serial_number_words[1];`

File: sht3x/sht3x.c (hunk 1 above)
Comment (abrauchli) [reply sht3x/sht3x.c]: that works for one word, but it's two words, so it'd need two more shifts. I'd rather stick to `be32_to_cpu` which should make everything above clear(er)

File: shtc1/shtc1.c (hunk 3 above)
Comment (abrauchli) [reply shtc1/shtc1.c]: yup

File: shtc1/shtc1.c (hunk 4 above)
Comment (abrauchli) [reply shtc1/shtc1.c]: meh, see comment above

**Comment by abrauchli** [comment]
@rnestler  The Serial Numbers now match ControlCenter

**Pull request merged**